filecoin-proofs/src/api/mod.rs

```diff
@@ -1,409 +1,167 @@
-                position: leaf_position,
-                number_of_leaves,
-                left_bytes: padded_left_bytes,
-                rest_bytes: padded_piece_length + padded_right_bytes,
+    // Holds intermediary pieces
+    let mut stack: Vec<PieceInfo> = vec![];
+
+    let mut current: PieceInfo = piece_infos.first().unwrap().clone();
+    // println!("current <- 0 ({})", u64::from(current.size) / 127);
+
+    for (_i, piece) in piece_infos.iter().skip(1).enumerate() {
+        // let i = i + 1;
+        use std::cmp::Ordering;
+
+        match piece.size.cmp(&current.size) {
+            Ordering::Equal => {
+                // println!(
+                //     "current <- H(current({}) | {}({}))",
+                //     i,
+                //     u64::from(current.size) / 127,
+                //     u64::from(piece.size) / 127
+                // );
+                current.size = UnpaddedBytesAmount(u64::from(current.size) + u64::from(piece.size));
+                let h = piece_hash(&current.commitment, &piece.commitment);
+                current.commitment.copy_from_slice(h.as_ref());
             }
-        })
-        .collect()
-}
-
-fn generate_piece_specs_from_source(
-    source: &mut (impl Read + Seek),
-    piece_lengths: &[UnpaddedBytesAmount],
-) -> error::Result<Vec<PieceSpec>> {
-    let pseudo_piece_specs = generate_pseudo_piece_specs(piece_lengths);
-
-    let mut piece_specs: Vec<PieceSpec> = vec![];
-
-    for pseudo_piece_spec in pseudo_piece_specs {
-        source.seek(SeekFrom::Current(
-            usize::from(pseudo_piece_spec.left_bytes) as i64
-        ))?;
-
-        let mut buf = vec![0; usize::from(pseudo_piece_spec.rest_bytes)];
-        source.read_exact(&mut buf)?;
-
-        let mut source = Cursor::new(&buf);
-        let comm_p = generate_piece_commitment_bytes_from_source::<PedersenHasher>(&mut source)?;
-
-        piece_specs.push(PieceSpec {
-            comm_p,
-            position: pseudo_piece_spec.position,
-            number_of_leaves: pseudo_piece_spec.number_of_leaves,
-        });
+            Ordering::Less => {
+                // println!("\t#S2 = {}", stack.len());
+                if stack.is_empty() {
+                    // println!("S2 <- {}({})", i, u64::from(piece.size) / 127,);
+                    stack.push(piece.clone());
+                } else if stack.last().unwrap().size == piece.size {
+                    // println!(
+                    //     "S2 <- H(s({}) | {} ({}))",
+                    //     u64::from(stack[stack.len() - 1].size) / 127,
+                    //     i,
+                    //     u64::from(piece.size) / 127
+                    // );
+                    {
+                        let l = stack.len();
+                        let el = &mut stack[l - 1];
+                        el.size = UnpaddedBytesAmount(u64::from(el.size) + u64::from(piece.size));
+                        let h = piece_hash(&el.commitment, &piece.commitment);
```

Comment:
I don't think it's guaranteed that the piece being processed will be the same size as the one on the top of the stack. If they aren't the same size, then we cannot hash them like this.

One way to guard against this risk would be to modify `piece-hash` to take `PieceInfo`s as input and refuse to hash mismatched pieces.

I think your algorithm will still do the right thing if you just remove this hashing entirely, and leave it to be performed by the reduction if appropriate.

Reply:
Looking at this again later, I think misread the brackets above. I think you do have that guarantee.